File: lc_soln/0199_binary_tree_right_side_view/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::vector<int> result;
        if (root == nullptr) {
            return result;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            while (n-- > 0) {
                TreeNode* node = q.front();
                q.pop();
                if (n == 0) {
                    result.push_back(node->val);
                }
                if (node->left != nullptr) {
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0199_binary_tree_right_side_view/solution.rs
struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = Vec::new();
        if let Some(r) = root {
            let mut q = VecDeque::new();
            q.push_back(r);

            while !q.is_empty() {
                let n = q.len();
                for i in 0..n {
                    if let Some(node) = q.pop_front() {
                        let node = node.borrow();
                        if i == n - 1 {
                            result.push(node.val);
                        }
                        if let Some(left) = &node.left {
                            q.push_back(left.clone());
                        }
                        if let Some(right) = &node.right {
                            q.push_back(right.clone());
                        }
                    }
                }
            }
        }
        result
    }
}

fn main() {}
